### DISCUSSION TITLE
Supabase Assistant feedback

### DISCUSSION THREAD

**Discussion by saltcod** [supabase/supabase#19594] [General]
**Please see updated discussion for v2 feedback:** 
https://github.com/orgs/supabase/discussions/19594

You've been asking for a better way to create RLS policies for a while now. During LWX in December 2023, we're happy to release Supabase Assistant. This tool brings the text-to-sql feature inside the SQL Editor to the RLS editor. We think AI can really help speed up the process of creating security policies. We'd love to know what you think. 

What do you like most about it? Least?
How are the AI responses? 
Are there any kinds of prompts that the Assistant struggles with? 
Anything you think could be improved? 
Anything you think is missing?

**Comment by GaryAustin1** [comment]
The assistant is not very tuned and somewhat random in approach to RLS including one bad policy for a simple task.
The task to change the policy from public to authenticated user only.
The best and correct answer is TO authenticated USING (true)

"allow only authenticated users to select"
Returns a bad policy that does not run at all.  

![badrls](https://github.com/supabase/supabase/assets/54564956/f189c35f-9f89-467e-9bc9-2eadfcb2763b)

"allow select only to authenticated users"
This policy at least works, but is a terrible way to do it for performance. 

![image](https://github.com/supabase/supabase/assets/54564956/aac632a4-a219-46bf-825e-ef3c65faea4e)


"allow only authorized users to select"
It made up a table that does not exist....
![anotherrls](https://github.com/supabase/supabase/assets/54564956/4bc93c07-0a9a-45f0-acdd-9d21be1cd4df)


This is similar query with a different answer that will run successfully, but would be very bad for security if the user does not know the TO field to spot the error.  I guess it could be argued that because I used authorized instead of authenticated any user with an api key is authorized.  
![image](https://github.com/supabase/supabase/assets/54564956/d0b07f9f-6f45-4487-86ac-ce57f67f4c2c)

**Reply by saltcod** [reply]
Thanks Gary — great feedback.  Out of curiosity — under AI Settings there, are you opted in or out here? 

<img width="754" alt="screenshot-2023-12-11-at-15 52 06" src="https://github.com/supabase/supabase/assets/105593/b17f7abe-9b67-4d63-a67c-1b45effa1cdf">

**Reply by GaryAustin1** [reply]
I thought it was on from the SQL AI, but it was off.  Not sure if it got reset with new stuff, or at some point I turned it off after testing the SQL AI.   That should fix the odd table it came up with.   

Might want to have some indicator about that (or maybe I should read instructions....)
I'll recheck the others and report back if it changes, those, but not sure why it would.

**Reply by GaryAustin1** [reply]
Ha.  Seems to like this bad method...
![image](https://github.com/supabase/supabase/assets/54564956/6069d651-8c3c-4858-92d0-49a3fa47f268)

**Reply by GaryAustin1** [reply]
Actually seems to do better if you use authenticated now.  So I guess it did not know the postgres roles either.  It got the TO correct on tests with the exact role name.

So you really should have that setting on, or it is pretty useless on RLS.

Edit:  Or not.   Seems random, but did the TO at least once.

![image](https://github.com/supabase/supabase/assets/54564956/c7b9abd3-5e98-47a8-a5be-11d64b05e3b3)

**Comment by saltcod** [comment]
A couple of users saying their schema is too large: 

Similar to this one:
https://github.com/orgs/supabase/discussions/16401

**Comment by galacticgibbon** [comment]
I didn't try it very long, I just found it less usable than the existing solution. I would like it more if the AI assistant suggested the rules within the existing editor, so that the table name is already defined. Becuase then I had to ask 'for X table, create an rls policy for...' which is cumbersome for long table names.

I've only ever made RLS policies through your editor so I'm not sure how much boilerplate you add to the query, and I like that I don't have to worry about that

**Reply by saltcod** [reply]
Great feedback, thanks @hamishjohnson

**Reply by shane-smith-1** [reply]
Yeah, second this. I really want some AI to help as I also have little RLS experiences (and SQL writing for that matter). Even simple stuff like changing ALL SELECT DELETE w/o needing to delete and re-add a query. (Or writing the correct query to make that change yourself).

Related feature request. I find myself pasting in RLS policies into an LLM and asking for anything im missing, any loopholes, vulternabilities etc. Could be nice to have some AI "'checking"" of policies. I understand its tricky to know peoples policy goals/needs though.

Thanks

**Comment by ste00martin** [comment]
basic things like asking for how to add RLS based on an authenticated users' phone number doesn't work.

**Comment by aaa3334** [comment]
Just had a look at this.
1) It is impossible to just edit or slightly adjust a current policy when you have this enabled here.. have do disable this.
2) It does not seem to generate anything useful...

I would much rather something which could auto copy and paste from one table to another (not even duplicating a table does this).

Seems it is still faster to copy from one table (manually) and add to the next. Perhaps it would be better either so we can set our own defaults (with a variable column which is for the current table we are adding rls to) or that we can 'copy' from another and use the same format (and auto replace the table name with the new one and provide it for review).
Could work with ai but the ai would then need to have access to the other ones which I dont think it does...

**Comment by benrobertsonio** [comment]
everything I generated with it was invalid SQL

**Comment by Pabluor** [comment]
Most of the time I try  to generate or update an existing policy with the AI assistant it does not work, always have errors. Trying to make the assistant handling the errors but does not work either. Had to go back to the new old way: chatgpt

**Comment by sapoepsilon** [comment]
5 out of 5 times did not work

**Comment by saltcod** [comment]
Closing this in favour of a new iteration: https://github.com/orgs/supabase/discussions/19594
Thanks to all for your feedback to get us to here!

**Comment by tecoad** [comment]
I have enable this feature but want to roll back. How to disable it?

**Comment by smartsense-as** [comment]
I’m sorry but the AI feature in supabase is totally useless. I think it would be better if you removed it than just have it there for the sake of jumping on the bandwagon! Very disappointing compared to some databade AI tools out there.
